jaxopt/_src/gauss_newton.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""GaussNewton algorithm in JAX."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+
+from dataclasses import dataclass
+
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src import implicit_diff as idf
+from jaxopt._src import loop
+from jaxopt._src import linear_solve
+
+
+class GaussNewtonState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  residual: Any
+  delta: Any
+  error: float
+  gradient: Any
+  jtj: Optional[Any] = None
```

Comment:
Why do you store `residual`, `delta`, `gradient` and `jtj`? Are they used in underlying applications?

Reply:
Yes, exactly, they are useful for interpreting the results for the underlying application. for example, if needed, eigenvalues of JT.J can be evaluated. For LM specifically, grad and delta can be useful to determine what cause the algorithm to stop/converge.